fhir-server/src/main/java/au/csiro/pathling/QueryHelpers.java

```diff
@@ -272,32 +272,39 @@ private static boolean hasColumn(@Nonnull final Dataset<Row> dataset,
     Dataset<Row> dataset = left.getDataset();
     final List<FhirPath> joinTargets = fhirPaths.subList(1, fhirPaths.size());
 
+    boolean nonLiteralEncountered = left instanceof NonLiteralPath;
     for (final FhirPath right : joinTargets) {
-      final List<Column> leftColumns = new ArrayList<>();
-      final List<Column> rightColumns = new ArrayList<>();
-      leftColumns.add(left.getIdColumn());
-      rightColumns.add(right.getIdColumn());
-      // If a $this context is present (i.e. the join is being performed within function arguments,
-      // with an item from the input collection as context), then we need to add the identity of the
-      // input element to the join. This is to prevent too many rows being generated when there are
-      // more rows than resources on either side of the join.
-      if (parserContext.getThisContext().isPresent()
-          && parserContext.getThisContext().get() instanceof NonLiteralPath
-          && left instanceof NonLiteralPath
-          && right instanceof NonLiteralPath) {
-        final NonLiteralPath nonLiteralThis = (NonLiteralPath) parserContext.getThisContext().get();
-        if (nonLiteralThis.getEidColumn().isPresent()) {
-          final Column thisEidColumn = nonLiteralThis.getEidColumn().get();
-          // Only use the element ID for the join if both datasets have a $this column, which means
-          // that they both originate from $this.
-          if (hasColumn(left.getDataset(), thisEidColumn)
-              && hasColumn(right.getDataset(), thisEidColumn)) {
-            leftColumns.add(thisEidColumn);
-            rightColumns.add(thisEidColumn);
+      if (right instanceof NonLiteralPath) {
+        if (nonLiteralEncountered) {
+          final List<Column> leftColumns = new ArrayList<>();
+          final List<Column> rightColumns = new ArrayList<>();
+          leftColumns.add(left.getIdColumn());
```

Comment:
I am not sure if this is OK given that left may be a Literal and not be part of any of the previous joins.

Reply:
Fixed in 59295803b97509c48ae9620dc3f13809c425b80a.